File: benchmarks/livermore_canonical/k19_linear_recurrence.f
C     Kernel 19 -- General Linear Recurrence Equations (canonical)
C     From "The Livermore Fortran Kernels" (UCRL-53745), F.H. McMahon, 1986
C     Canonical body (forward + backward sweep, both executed):
C       KB5I = 0
C       DO 191 k=1,n: B5(k+KB5I) = SA(k) + STB5*SB(k); STB5 = B5(k+KB5I) - STB5
C       DO 193 i=1,n: k=n-i+1; B5(k+KB5I) = SA(k) + STB5*SB(k); STB5 = B5(k+KB5I) - STB5
C
      PROGRAM K19
      IMPLICIT DOUBLE PRECISION (A-H,O-Z)
      INTEGER REP
      DIMENSION B5(101), SA(101), SB(101)
      DIMENSION SPACER(39)
C
C     Initialize SPACER and pull STB5 from canonical position 35
C
      FUZZ = 1.2345D-3
      BUZZ = 1.0D0 + FUZZ
      FIZZ = 1.1D0 * FUZZ
      DO 5 K = 1, 39
         BUZZ = (1.0D0 - FUZZ) * BUZZ + FUZZ
         FUZZ = -FUZZ
         SPACER(K) = (BUZZ - FIZZ) * 0.1D0
    5 CONTINUE
      STB5_INIT = SPACER(35)
C
C     Initialize SA using SIGNEL
C
      FUZZ = 1.2345D-3
      BUZZ = 1.0D0 + FUZZ
      FIZZ = 1.1D0 * FUZZ
      DO 10 K = 1, 101
         BUZZ = (1.0D0 - FUZZ) * BUZZ + FUZZ
         FUZZ = -FUZZ
         SA(K) = (BUZZ - FIZZ) * 0.1D0
   10 CONTINUE
C
C     Initialize SB using SIGNEL
C
      FUZZ = 1.2345D-3
      BUZZ = 1.0D0 + FUZZ
      FIZZ = 1.1D0 * FUZZ
      DO 15 K = 1, 101
         BUZZ = (1.0D0 - FUZZ) * BUZZ + FUZZ
         FUZZ = -FUZZ
         SB(K) = (BUZZ - FIZZ) * 0.1D0
   15 CONTINUE
C
C     Zero out B5
C
      DO 20 K = 1, 101
         B5(K) = 0.0D0
   20 CONTINUE
C
      N = 101
C
C     Kernel loop (canonical: forward sweep then backward sweep)
C
      DO 500 REP = 1, 13000000
         STB5 = STB5_INIT
         KB5I = 0
C        Forward sweep (DO 191)
         DO 200 K = 1, N
            B5(K+KB5I) = SA(K) + STB5*SB(K)
            STB5 = B5(K+KB5I) - STB5
  200    CONTINUE
C        Backward sweep (DO 193)
         DO 300 I = 1, N
            K = N - I + 1
            B5(K+KB5I) = SA(K) + STB5*SB(K)
            STB5 = B5(K+KB5I) - STB5
  300    CONTINUE
  500 CONTINUE
C
      WRITE(*,900) B5(N)
  900 FORMAT('b5(n) = ', E25.15)
      END

File: benchmarks/livermore_canonical/k19_linear_recurrence.c
/* K19 -- General Linear Recurrence Equations (canonical, 1-based indexing)
   Canonical body (forward + backward sweep, both executed):
     KB5I = 0
     DO 191 k=1,n: B5(k+KB5I) = SA(k) + STB5*SB(k); STB5 = B5(k+KB5I) - STB5
     DO 193 i=1,n: k=n-i+1; B5(k+KB5I) = SA(k) + STB5*SB(k); STB5 = B5(k+KB5I) - STB5 */
#include <stdio.h>
#include <time.h>
#include "signel.h"

#define N     101
#define NREPS 13000000

int main(void) {
    double b5[102], sa[102], sb[102];
    double spacer[40];

    signel(spacer+1, 39);
    double stb5_init = spacer[35];

    signel(sa+1, 101);
    signel(sb+1, 101);
    for (long k = 1; k <= 101; k++) b5[k] = 0.0;

    struct timespec t0, t1;
    clock_gettime(CLOCK_MONOTONIC, &t0);

    long kb5i;
    double stb5;
    for (long rep = 0; rep < NREPS; rep++) {
        stb5 = stb5_init;
        kb5i = 0;
        /* Forward sweep */
        for (long k = 1; k <= N; k++) {
            b5[k+kb5i] = sa[k] + stb5 * sb[k];
            stb5 = b5[k+kb5i] - stb5;
        }
        /* Backward sweep */
        for (long i = 1; i <= N; i++) {
            long k = N - i + 1;
            b5[k+kb5i] = sa[k] + stb5 * sb[k];
            stb5 = b5[k+kb5i] - stb5;
        }
    }

    clock_gettime(CLOCK_MONOTONIC, &t1);
    double elapsed = (t1.tv_sec - t0.tv_sec) + (t1.tv_nsec - t0.tv_nsec) * 1e-9;
    printf("elapsed: %.6f s\n", elapsed);
    printf("b5[n] = %f\n", b5[N]);
    return 0;
}
